rust/lance/src/dataset/rowids.rs

```diff
@@ -3,15 +3,70 @@
 
 use super::Dataset;
 use crate::{Error, Result};
-use futures::{StreamExt, TryStreamExt};
+use futures::{Stream, StreamExt, TryFutureExt, TryStreamExt};
 use snafu::{location, Location};
 use std::sync::Arc;
 
 use lance_table::{
-    format::RowIdMeta,
-    rowids::{read_row_ids, RowIdIndex},
+    format::{Fragment, RowIdMeta},
+    rowids::{read_row_ids, RowIdIndex, RowIdSequence},
 };
 
+/// Load a row id sequence from the given dataset and fragment.
+pub async fn load_row_id_sequence(
+    dataset: &Dataset,
+    fragment: &Fragment,
+) -> Result<Arc<RowIdSequence>> {
+    // Virtual path to prevent collisions in the cache.
+    let path = dataset.base.child(fragment.id.to_string()).child("row_ids");
+    match &fragment.row_id_meta {
+        None => Err(Error::Internal {
+            message: "Missing row id meta".into(),
+            location: location!(),
+        }),
+        Some(RowIdMeta::Inline(data)) => {
+            dataset
+                .session
+                .file_metadata_cache
+                .get_or_insert(&path, |_path| async { read_row_ids(data) })
+                .await
+        }
+        Some(RowIdMeta::External(file_slice)) => {
+            dataset
+                .session
+                .file_metadata_cache
+                .get_or_insert(&path, |_path| async {
+                    let path = dataset.base.child(file_slice.path.as_str());
+                    let range = file_slice.offset as usize
+                        ..(file_slice.offset as usize + file_slice.size as usize);
+                    let data = dataset
+                        .object_store
+                        .open(&path)
+                        .await?
+                        .get_range(range)
+                        .await?;
+                    read_row_ids(&data)
+                })
+                .await
+        }
+    }
+}
+
+/// Load row id sequences from the given dataset and fragments.
+///
+/// Returned as a vector of (fragment_id, sequence) pairs. These are not
+/// guaranteed to be in the same order as the input fragments.
+pub fn load_row_id_sequences<'a>(
+    dataset: &'a Dataset,
+    fragments: &'a [Fragment],
+) -> impl Stream<Item = Result<(u32, Arc<RowIdSequence>)>> + 'a {
```

Comment:
Minor nit: do you anticipate using this in places where you don't want to immediately call `try_collect`?  Otherwise you could make the method async and just return `Vec` for a simpler interface.